lib/rules/deprecations/base-deprecation.js

```diff
@@ -12,8 +12,7 @@ class BaseDeprecation extends BaseChecker {
     const pragma = node.name
     const value = node.value
     if (pragma === 'solidity') {
-      this.active =
-        config.compareSemVer(value.replace(/[^0-9.]/g, ''), this.deprecationVersion()) >= 0
+      this.active = semver.gte(value.replace(/[^0-9.]/g, ''), this.deprecationVersion())
```

Comment:
Hey @junderw, sorry for not having reviewed this yet. I will make a proper review as soon as I have some time.

For the time being, I can say that this line doesn't feel quite right. For example, this would fail if the version range is `<0.7.0`.

Actually, the more I think about this, the more I feel the deprecation mechanism doesn't make sense, because it's impossible to know which version of the compiler is actually being used. For example, let's say there's a rule that is invalid for the range `<0.7.0`, and a file has a version range `>=0.6.0`. The rule shouldn't be deprecate when using the version `0.7.0` of the compiler, but there's no way to know this from the pragma alone.

I need to analyze this more in depth. But my first instinct would be to let the user handle the problem however they see fit. For example, if someone is using 0.7.0 they should disable this rule, or add some option. Maybe instead of invalidating rules for some versions, the best approach is to give better mechanisms for users to configure the rules (e.g. disabling some rule in a directory, idk). Again, not really sure, I will think about this a little more.

Reply:
Sorry for the confusion.

The BaseDeprecation change was an additional cleanup and does not change any behavior. I can remove it if necessary, I just thought it should be consistent in how it deals with semver comparisons.

Timeline:

1. I was looking for examples where this library handles semver checking and found basedeprecation.
2. The semver compare that existed was lacking, so I re-implemented it as a util and used it in my change AND basedeprecation.
3. You suggested semver package, so I changed BOTH to use semver.

Reply:
re: how to handle it.

I don't know how you can get around semver gating it...

0.7.0 forbids it but 0.6.12 requires it.

No matter what error or warning we make, it will be causing an endless loop for one of those two versions of fixing compiler then fixing lint until an ignore comment is added.

Reply:
Yes, this is totally my fault! I should've realized before that this deprecation thing doesn't really work :sweat_smile:

Do you mind rolling back all the deprecation stuff, and just leaving the updating of the rule? Sorry for all this back and forth.

Reply:
Fixed.